### PR TITLE
Generic errors for entire crate

### PR DESCRIPTION
**Pull request by landesfeind**
This branch  tackles the issue #215 to provide global errors

### CODE REVIEW

File: src/errors.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    #[error("unable to open SAM/BAM/CRAM index for {target}")]
+    BamInvalidIndex { target: String },
+    #[error("failed to write record (out of disk space?)")]
+    Write,
+    #[error("invalid record in SAM/BAM/CRAM file")]
+    BamInvalidRecord,
+    #[error("truncated record in SAM/BAM/CRAM file")]
+    BamTruncatedRecord,
+    #[error("format of SAM files are not indexable")]
+    NotIndexable,
```
Comment (dlaehnemann) [inline src/errors.rs:60]: I found this wording ambiguous and looked into the htslib code. I would suggest the following wording:
```suggestion
    #[error("format not indexable by htslib (format is detected as something else than SAM/BAM/CRAM)")]
    NotIndexable,
```
The respective code that can produce the return value of `-3` that is this error uses, is here:
https://github.com/samtools/htslib/blob/34ba6c726925efceaa2cc995d7d4b0409907f331/sam.c#L840-L880

**Review by dlaehnemann — COMMENTED** [review]
Thanks a lot for getting this started!

And regarding your question over in issue #215 , I would intuitively prefer to only have general errors in `src/errors.rs` and module-specific errors in something like `src/bam/errors.rs`. This should then of course be documented accordingly.

But I also see the advantage of having all errors in one place, because it avoids that you always have to look in two places for available errors (i.e. in `src/errors.rs` and `/src/bam/errors` for the `bam` module). But I think if it is clear from the docs (both the general and the module docs), that these two possible locations exist for any module, this should be fine, right?

**Comment by dlaehnemann** [comment]
Should be update to include the errors added with PR #258 .

File: src/errors.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    // Errors for faidx
+    #[error("The given position is too large to be converted to i64")]
+    PositionTooLarge,
+
+    // Errors for Tbx
+    #[error("previous iterator generation failed")]
+    NoIter,
+    #[error("truncated tabix record")]
+    TabixTruncatedRecord,
+    #[error("invTabixalid tabix index")]
```
[the remark this answers is not in the document]
Comment (landesfeind) [reply src/errors.rs]: Solved in 14fae4d2322e5f83fbede13b0961521b1949ea21

File: src/errors.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    NonUnicodePath,
+    #[error("failed to fetch region")]
+    Fetch,
+    #[error("error seeking to file offset")]
+    FileSeek,
+    #[error("error seeking to {contig:?}:{start} in indexed file")]
+    GenomicSeek { contig: String, start: u64 },
+    #[error("sequence {sequence} not found in index")]
+    UnknownSequence { sequence: String },
+    #[error("error setting threads for for file reading")]
```
[the remark this answers is not in the document]
Comment (landesfeind) [reply src/errors.rs]: Solved in 14fae4d2322e5f83fbede13b0961521b1949ea21

File: src/errors.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    SetThreads,
+
```
[the remark this answers is not in the document]
Comment (landesfeind) [reply src/errors.rs:27]: Solved in d1ac39f82c8c70bf695fcc1f57f1982e9c28a6b6

File: src/errors.rs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    #[error("invalid reference path {path}")]
+    InvalidReferencePath { path: PathBuf },
+    #[error("invalid compression level {level}")]
+    InvalidCompressionLevel { level: u32 },
+    #[error("unable to open SAM/BAM/CRAM file at {target}")]
+    BamOpen { target: String },
+    #[error("unable to open SAM/BAM/CRAM index for {target}")]
+    BamInvalidIndex { target: String },
+    #[error("failed to write record (out of disk space?)")]
+    Write,
```
[the remark this answers is not in the document]
Comment (landesfeind) [reply src/errors.rs]: Solved in 549525fdcc8d697bebca2697ce3103d5c33d5a7c

File: src/errors.rs (hunk 1 above)
Comment (landesfeind) [reply src/errors.rs]: Solved in ba84f2c2ff670f4be072ebd4b8c467e9b850737b

File: src/errors.rs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    // Errors for BAM
+    #[error("error parsing CIGAR string: {msg}")]
+    ParseCigar { msg: String },
+    #[error("unexpected CIGAR operation: {msg}")]
+    UnexpectedCigarOperation { msg: String },
+    #[error("error parsing SAM record: {rec}")]
+    ParseSAM { rec: String },
+    #[error("invalid reference path {path}")]
+    InvalidReferencePath { path: PathBuf },
```
[the remark this answers is not in the document]
Comment (landesfeind) [reply src/errors.rs:51]: Solve by ba84f2c2ff670f4be072ebd4b8c467e9b850737b

**Comment by landesfeind** [comment]
Thank you for the review. I resolved the threads covering typos and structure.

Regarding the topic of one-error-enum, I vote pro one-error-enum, as long as it is not too many errors (my magical number here tends toward 30 specific errors). 

This makes working with the library very convenient because when you wrap around it, you can only define one error, like (in `my-crate-using-htslib/src/errors.rs`):
```rust
pub enum MyErrors {
   ...,
   HtslibError { e : rust_htslib::Error }
}
```

Also working with errors inside of rust-htslib is simpler. If we use generic and specific errors, we would need a pattern like (e.g., in `bam/errors.rs`):
```rust
pub enum BamError {
  ...,
  General { e: crate::error::GeneralError }
}
``` 
Not sure if this is intuitive.

Last, there would always be discussions about "when is a error generic enough be be put in the crate-wide errors?" :-)

**Review by dlaehnemann — COMMENTED** [review]
Thanks for addressing all the little comments!

I do see point of having all errors in one place and having to only import one error enum in general. In case we go for this, I would opt for a Format-specific prefix for each error (i.e. `BamSomethingError` for all `Bam` errors, `TbxSomethingError` for all Tabix errors, ...). This makes it easy to find the available errors for a format via autocompletion in an IDE.

Regarding the wrapping when used in another crate, I am not sure. Wouldn't most errors simply be thrown by rust_htslib itself when calling one of its functions? And for other cases, where you want to use an error in a function from that new crate, wouldn't you just import the enum and use it? So a bam module error would look something like:
```
use rust_htslib
fn do_something(input1: i1) -> Result<> {
    match i1 {
        TypeX(x) => Ok(x),
        _ => rust_htslib::bam::SomeError
    }
}
```
But maybe it's just a use case I am not familiar with. And in any case, I think having all errors in one place and prefixing them should work.

The only drawback I could see would be if we later on start introducing features to speed up compilation when you only want to use certain modules of the crate (something like this is being discussed for rust-bio [here](https://github.com/rust-bio/rust-bio/issues/342)). In this case, you would be dragging errors along unnecessarily. But at the same time, that doesn't sound like a great burden -- it's only a bunch of entries in an enum, after all...

What do you think?

File: src/bam/mod.rs [hunk 7]
```diff
@@ -24,11 +23,12 @@ use std::str;
 
 use url::Url;
 
+use crate::errors::*;
```
[the remark this answers is not in the document]
Comment (landesfeind) [reply src/bam/mod.rs]: Solved in e57f9a700bc9c964ed3226fb6ca98a795a7b763b

**Comment by landesfeind** [comment]
Having prefixed errors for specific sub-modules is a very good idea and I implemented it in ef894c04a78c04daf0f5ed68598f779ee3a02bf1

Regarding crate- versus module-wide errors, I think crate-wide errors are handy whenever you write a function that uses multiple sub-modules of htslib.

Let's assume we write a small tool (or even a library) that iterates over a VCF/BCF to count the mutational signature of the sample by means of how often does which mutation occur. The tool is not interested in single SNPs but in the genomic triplet change (1 base up- and down-stream). We want to write a function that interprets the VCF record and returns the genomic triplet and alternative. The function therefore uses bcf- and faidx-modules. 

In example 1 below, the function has a rather simple return type based on the rust-htslib error/result. For the second example, in which we have module-specific errors, the returns type must be wrapped because two error-enums can be returned. 

The example may seem a bit artificial but it's actually derived from a real-world issue I had. I hope it illustrates my point even though I am aware that there may be other ways how to implement this cleaner or smarter ;-)

**Example 1 - One error for whole crate**
```rust
fn get_mutation_codon(vcf_record: vcf::Record, faidx: faidx::Reader) -> rust_htslib::error::Result<(String, String)> {
	// First process the vcf and do something that could result in an error 
	...

	// Then read the Faidx to retrieve genomic sequence (this could also result in an error)
	...

	(genomic_sequence, vcf_alternative_allele)
}
```

**Example 2 - Errors for specific sub-modules**
```rust
enum GetMutationCodonError {
	HtslibBcfError { e : rust_htslib::bcf::Error }
	HtslibFaidxError { e : rust_htslib::faidx::Error }
}

fn get_mutation_codon(vcf_record: vcf::Record, faidx: faidx::Reader) -> rust_htslib::error::Result<(String, String)> {
	// First process the vcf and do something that could result in an error 
	...

	// Then read the Faidx to retrieve genomic sequence (this could also result in an error)
	...

	(genomic_sequence, vcf_alternative_allele)
}
```


# Enabling features

Also this should be possible. I made a simple test and the following compiles - even though I don't know how it affects compilation speed. 
```rust
pub enum Error {
    ...
    #[error("failed to remove alleles in BCF/VCF record")]
    BcfRemoveAlleles,

		#[cfg(feature = "with_specific_submodule")]
		SubModuleError1
}
```

**Review by dlaehnemann — APPROVED** [review]
@landesfeind Thanks a lot for taking the time to pen down this example (in Example 2, I think you meant for the return type to be something like `GetMutationCodonError`, right?).
In any case, this makes sense, now. And all the better, if feature gatewaying also works on the level of individual errors in the enum. So I think we are good to go, here.

One last thing, which I also always keep on forgetting: Would you write a quick note into the `CHANGELOG.md` for the next release. Just makes creating those releases a tad bit faster...

**Comment by landesfeind** [comment]
> ... (in Example 2, I think you meant for the return type to be something like `GetMutationCodonError`, right?).

Oh. Yes you are right. I should always doc-test ;-p

> One last thing, which I also always keep on forgetting: Would you write a quick note into the `CHANGELOG.md` for the next release. Just makes creating those releases a tad bit faster...

Done in 860384c2622081d93ea557a3417041bea0594b57 !

Last suggestion/question: should we `pub use errors::*;` in `lib.rs` to export the errors directly on highest crate-level? But that is totally cosmetics :-)

**Comment by dlaehnemann** [comment]
I'd say including them via `use rust_htslib::errors::Error` works well and makes it clear where to find them in the hierarchy. So no need to explicitly export at the highest level. But as above, I might not be seeing some important application...?!

**Comment by landesfeind** [comment]
It would only be "cosmetics" (one `::errors` less, no algorithmic improvement). Lets keep them where they are :-)

**Pull request merged**